javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+            Objects.requireNonNull(predicate, "predicate is null");
+            return iterator().span(predicate).map(this::createFromAll, this::createFromAll);
+        }
+
+        @Override
+        public BitSet<T> scan(T zero, BiFunction<? super T, ? super T, ? extends T> operation) {
+            Objects.requireNonNull(operation, "operation is null");
+            return Collections.scanLeft(this, zero, operation, new java.util.ArrayList<T>(), (arr, t) -> {
+                arr.add(t);
+                return arr;
+            }, this::createFromAll);
+        }
+
+        @Override
+        public Tuple2<BitSet<T>, BitSet<T>> partition(Predicate<? super T> predicate) {
+            Objects.requireNonNull(predicate, "predicate is null");
+            return iterator().partition(predicate).map(this::createFromAll, this::createFromAll);
+        }
+
+        @Override
+        public BitSet<T> filter(Predicate<? super T> predicate) {
+            Objects.requireNonNull(predicate, "predicate is null");
+            final BitSet<T> bitSet = createFromAll(iterator().filter(predicate));
+            return (bitSet.length() == length()) ? this : bitSet;
+        }
+
+        @Override
+        public  <C> Map<C, BitSet<T>> groupBy(Function<? super T, ? extends C> classifier) {
+            Objects.requireNonNull(classifier, "classifier is null");
+            return iterator().groupBy(classifier).map((key, iterator) -> Tuple.of(key, createFromAll(iterator)));
+        }
+
+        @Override
+        public Comparator<T> comparator() {
+            return (t1, t2) -> Integer.compare(toInt.apply(t1), toInt.apply(t2));
+        }
+
+        @Override
+        public BitSet<T> takeWhile(Predicate<? super T> predicate) {
+            Objects.requireNonNull(predicate, "predicate is null");
+            final BitSet<T> result = createFromAll(iterator().takeWhile(predicate));
+            return (result.length() == length()) ? this : result;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public BitSet<T> addAll(Iterable<? extends T> elements) {
+            final Stream<Integer> source = Stream.ofAll(elements).map(toInt);
+            final long[] copy = copyExpand(1 + (source.max().getOrElse(0) >> ADDRESS_BITS_PER_WORD));
+            source.forEach(element -> {
+                if (element < 0) {
+                    throw new IllegalArgumentException("bitset element must be >= 0");
+                }
+                setElement(copy, element);
+            });
+            return fromBitMaskNoCopy(copy);
+        }
+
+        @Override
+        public boolean contains(T t) {
+            final int element = toInt.apply(t);
+            if (element < 0) {
+                throw new IllegalArgumentException("bitset element must be >= 0");
+            }
+            final int index = element >> ADDRESS_BITS_PER_WORD;
+            return (getWord(index) & (1L << element)) != 0;
```

Comment:
could we extract all the `1L << element` to something like `private int indexOf(element)`?